src/Mvc/Mvc.Core/src/Routing/UrlHelper.cs

```diff
@@ -32,7 +32,21 @@ public UrlHelper(ActionContext actionContext)
         /// Gets the top-level <see cref="IRouter"/> associated with the current request. Generally an
         /// <see cref="IRouteCollection"/> implementation.
         /// </summary>
-        protected IRouter Router => ActionContext.RouteData.Routers[0];
+        protected IRouter Router
+        {
+            get
+            {
+                var routers = ActionContext.RouteData.Routers;
+                if (routers.Count == 0)
+                {
+                    throw new InvalidOperationException("Could not find an IRouter associated with the ActionContext. "
+                        + "If your application is using endpoint routing then you can get a UrlHelper instance with "
+                        + "dependency injection or use Microsoft.AspNetCore.Routing.LinkGenerator.");
```

Comment:
This is false. You can't get a `UrlHelper` from DI. That's one of the problems that link generator solves.

users can instead resolve an `IUrlHelperFactory` and put in the action context. Don't have an action context? that's another one of the problems link generator solves.